adopt shared header template on the dashboard view

PaarthAgarwal opened the pull request:
Addresses #8539.
Applied shared header to dashboard.
Before:

![Screenshot from 2022-06-30 15-57-54](https://user-images.githubusercontent.com/86092410/176655625-c2433e0b-9418-47da-a494-13defce6118b.png)


After:

![Screenshot from 2022-07-04 18-29-46](https://user-images.githubusercontent.com/86092410/177203944-802f8354-adec-4cb5-b24f-265bfc5bdfe4.png)

![Screenshot from 2022-07-04 18-31-07](https://user-images.githubusercontent.com/86092410/177203959-079c89b5-ca24-4016-9068-7669e3bcdf0f.png)

![Screenshot from 2022-07-04 18-31-35](https://user-images.githubusercontent.com/86092410/177203961-1dffd6fc-08c9-4647-a6db-6a446b13f63b.png)



 
Edit1: Updated Screenshot
Edit2: Updated Screenshot

lb- commented on wagtail/admin/templates/wagtailadmin/shared/header.html:
```diff
@@ -14,15 +14,19 @@
+                        {% icon class_name="w-header__icon" name=icon %}
+                    {% elif avatar %}
+                        <div class="avatar"><img src="{% avatar_url user %}" alt="" /></div>{% endif %}
                     {{ title }}{% if subtitle %} <span class="header__subtitle">{{ subtitle }}</span>{% endif %}
```
I recommend seeing how the username goes passed in to the `subtitle` instead. while this becomes part of the `h1` I think that kind of makes sense "Welcome to the blah blah LB Johnston".

lb- commented on wagtail/admin/templates/wagtailadmin/shared/header.html:
```diff
@@ -14,15 +14,19 @@
+                    {% if icon %}
+                        {% icon class_name="w-header__icon" name=icon %}
+                    {% elif avatar %}
+                        <div class="avatar"><img src="{% avatar_url user %}" alt="" /></div>{% endif %}
```
```suggestion
                        <div class="avatar"><img src="{% avatar_url user %}" alt="" /></div>
                    {% endif %}
```
Small nitpick but try to keep the closing endif on the same line indentation as the if opening.

lb- replied:
Remember to add that the template accepts an avatar url in the template description at the top.

lb- commented on wagtail/admin/templates/wagtailadmin/shared/header.html:
```diff
@@ -14,15 +14,19 @@
 
 {% endcomment %}
-<header class="header {% if merged %}merged{% endif %} {% if search_form %}hasform{% endif %}">
+<header class="header{% if merged %} merged{% endif %}{% if search_form %}hasform{% endif %} {{ classname }}">
```
Maybe avoid this unless you need it,
But if you do think we need this ability to add a classname please ensure you also add it to the description above (variables accepted by this template).

lb- replied:
Also be careful here, I think you accidentally broke the existing classes
`hasform` now shows bang up against `header` if used.

```suggestion
<header class="header {{ classname }} {{ merged|yesno:" merged ," }}{{ search_form|yesno:" hasform ," }}">
```

There might be a nicer way but you could leverage the `yesno` filter to wrap the class with spaces if it it is needed https://docs.djangoproject.com/en/4.0/ref/templates/builtins/#yesno


But longer term - we may need to find a better way for this.

lb- commented on wagtail/admin/templates/wagtailadmin/shared/header.html:
```diff
@@ -14,15 +14,19 @@
+                    {% if icon %}
+                        {% icon class_name="w-header__icon" name=icon %}
+                    {% elif avatar %}
+                        <div class="avatar"><img src="{% avatar_url user %}" alt="" /></div>{% endif %}
```
Be careful with the way you are reading the avatar here, you are reading it from context but not passing it in. instead the idea is that the `avatar` variable is the fully resolved URL.

`src="{{ avatar }}}`

Otherwise this template becomes more intertwined with requiring both the user and the template tag `avatar_url` in the context (this may also break storybook).

lb- commented on wagtail/admin/templates/wagtailadmin/home.html:
```diff
@@ -6,17 +6,8 @@
-        </div>
-    </header>
-
+    {% include "wagtailadmin/shared/header.html" with title=site_name avatar=1 username=user|user_display_name merged=1 %}
```
as noted below, it would be better to pass in the actual avatar URL to avoid the shared header template needing to have the overhead of using the template tag itself.

if `{% avatar_url user as avatar %}` does not work, you may need to pass it down from the template view.

```suggestion
    {% avatar_url user as avatar %}
    {% include "wagtailadmin/shared/header.html" with title=site_name avatar=avatar subtitle=user|user_display_name merged=1 %}
```

Also, as noted - try to keep the existing 'slots' where possible, this may look and work great with the subtitle being the username (but it may need some styling tweaks specific to this page)

lb- commented on client/scss/layouts/_home.scss:
```diff
@@ -1,31 +1,6 @@
-  }
+.homepage .user-name {
+  font-size: 1.3em;
+  font-weight: 600;
```
it would be nice to not need this styling either - if using the subtitle variant we may be fine OR we may need some of these styles still.

lb- commented on wagtail/admin/templates/wagtailadmin/shared/header.html:
```diff
@@ -14,15 +14,19 @@
+                        <div class="avatar"><img src="{% avatar_url user %}" alt="" /></div>{% endif %}
                     {{ title }}{% if subtitle %} <span class="header__subtitle">{{ subtitle }}</span>{% endif %}
                 </h1>
+                {%if username %}<div class="user-name">{{ username }}</div>{% endif %}
```
as per above, let's see if we can avoid adding this variable usage.

lb- reviewed (commented):
@PaarthAgarwal  great start on the way here, I would like to see the styling get a bit closer to the current state though.

I have put specific code feedback in the review comments, but also some general feedback

* The PR title could be a bit clearer 'use/adopt shared header template on the dashboard view'
* When the shared header template changes are closer, can you also include a storybook (pattern library) variant that has the avatar in it, this will really help for testing later

Great start here!

lb- commented on wagtail/admin/templates/wagtailadmin/home.html:
```diff
@@ -6,17 +6,8 @@
-        </div>
-    </header>
-
+    {% include "wagtailadmin/shared/header.html" with title=site_name avatar=1 username=user|user_display_name merged=1 %}
```
Also the title is wrong - it should be the result of 
`{% block branding_welcome %}{% blocktrans trimmed %}Welcome to the {{ site_name }} Wagtail CMS{% endblocktrans %}{% endblock %}`

But - you may need to set up this translation at the view level.

lb- replied:
ok, did a quick test and this should do the trick.

In the file ...`wagtail/admin/views/home.py`

```

    def get_context_data(self, **kwargs):
        context = super().get_context_data(**kwargs)
        panels = self.get_panels()
        site_details = self.get_site_details()

        context["media"] = self.get_media(panels)
        context["panels"] = sorted(panels, key=lambda p: p.order)
        context["user"] = self.request.user
        context["welcome_title"] = gettext_lazy("Welcome to the %(site_name)s Wagtail CMS") % { 'site_name': site_details['site_name'] } # add this line - it is the same translation block but done at the view level instead.

        return {**context, **site_details}
```

gasman replied:
Bear in mind that people routinely replace this text via template overrides - if we're moving this out of home.html, we'll need to provide an alternative way to customise it (and cover it in the release notes).

lb- replied:
Thanks Matt.
---
Do you know of a Django way to make a block become a value?

For example;
```
{% block branding_welcome as my_header_title %}{% endblock %}
```

Otherwise, not sure how we can make this easily overridable, but I will have a think;

* template inheritance (although, the current goal is to avoid that where possible as per parent issue)
* document that the home view can be extended, developers can customise the `get_context_data` and override the url with their custom home view (not sure if Django makes that easy though as the url name will exist twice).

lb- replied:
Really not sure where to head with this. I personally feel that inheritance here would be simpler and have more flexibility.

We could add a title block around the existing template and then use that to have the block behaviour. Developers will need to override the home page specific header template instead of the  home view. 

This will make it easier to align with how the ModelAdmin header template is built also. 

---

Another idea - what if we just put the welcome label in it's own template file? Or even don't create one and in our view's context we search for one and if not found fall back to the view default welcome.

---

I am concerned that if we try to force everything into component style template tags we may create something that is overall less flexible for developers. However, we do need to try to get the html content to be mostly consistent so that it is easier to do styling.  

But maybe it's not worth changing this? Just apply more styles on top for each version. 

@thibaudcolas please provide any additional ideas. 
@PaarthAgarwal have a read of all this and please add any thoughts you have.

lb- replied:
After some testing, I think the isolated welcome title template may be the simplest and supports the original flexibility, but will be a change for developers.

This 'pulls up' the content generation into the template view still instead of pushing down with layers of component/endcomponent templates.

---

Note: I tried a new `Component` but it felt like overkill for this.

## 1.  - create a new template file 

_Note: We could even preserve the block here so that developers have even less to change._

`wagtail/admin/templates/wagtailadmin/home/branding_welcome_title.html`
```html
{% load i18n %}
{% blocktrans trimmed %}Welcome to the {{ site_name }} Wagtail CMS{% endblocktrans %}
```

## 2. create a new method on the `HomeView` class

```python
    def get_welcome_title(self, context):
        return get_template("wagtailadmin/home/branding_welcome_title.html").render(context)
```

## 3. use this method in the `get_context_data`

```python
    def get_context_data(self, **kwargs):
        context = super().get_context_data(**kwargs)
        panels = self.get_panels()
        site_details = self.get_site_details()

        context["media"] = self.get_media(panels)
        context["panels"] = sorted(panels, key=lambda p: p.order)
        context["user"] = self.request.user

        context = {**context, **site_details} # set up the context earlier

        context["welcome_title"] = self.get_welcome_title(context) # add this line

        return context
```

## 4. update documentation

Note: This assumes we will not be reusing the block `branding_welcome`...

````markdown

### `branding_welcome`

```{versionchanged} 4.0
```

To replace the welcome message on the admin home page (dashboard), create a template file `dashboard/templates/wagtailadmin/home/branding_welcome_title.html` that overrides the content:

```html+django
Welcome to Frank's Site
```
````

5. Update the release notes with the change.

* ... can be done when merged.

thibaudcolas replied:
I think it’s time to introduce new template tags, so we don’t have to switch back and forth between Python and templates to do relatively basic composition of components. Clearly the vanilla `include` and `block` aren’t fit for purpose when componentising a UI like this – we need a way to pass template fragments to arbitrary components via composition rather than inheritance.

My preference would be to introduce generic tags that can support those types of patterns, rather than specialised tags for each component. At the moment I think the flexibility of the [`fragment` tag from Slippers](https://mitchel.me/slippers/docs/template-tags-filters/#fragment) would be the best fit for Wagtail. Assuming it supports rendering blocks, it’d look something like:

```html
{% fragment as site_name %}
  {% block branding_welcome %}{% blocktrans trimmed %}Welcome to the {{ site_name }} Wagtail CMS{% endblocktrans %}{% endblock %}
{% endfragment %}
{% include "wagtailadmin/shared/header.html" with title=site_name avatar=1 username=user|user_display_name merged=1 %}
```

See for example:

- https://mitchel.me/slippers/docs/template-tags-filters/#fragment

gasman replied:
I agree with @thibaudcolas - if the situation calls for being able to render a component by composing template fragments, we should adopt a solution that supports that, rather than trying to work around that requirement.

Having said that - if moving the title element to its own include is the simplest way to unblock this, then let's do that. While it is a breaking change for site implementers, I don't think there was an expectation that we'd keep that block in place in that particular template forever, and this change should be compatible with whatever mechanism we introduce in future.

lb- replied:
Thanks all. This is super great feedback. I'll walk through this on my Monday call with Paarth. 

If it's ok - we will probably opt for Matt's approval to unblock this with the simplest way forward (own include for title element). 

I agree that the fragment concept and also the ongoing discussion about a component type template tag is super important. 

We may end up writing this again in a few weeks (depending on the capacity for the team to also bring in these new template tag systems) which is fine. In the short term for 4.0.RC looming, we are hoping for some quicker visual wins. 

I understand we need to be careful to consider existing use cases for developers though and will be more vigilant in assessing existing block usage.

lb- commented:
@PaarthAgarwal Can you please do a seperate branch / PR for the dashboard stats design updates. 

It will.be easier to review in isolation. 

Plus - we may end up getting the stats thing done much easier due to the discussion about header component stuff that is ongoing.

PaarthAgarwal commented:
> @PaarthAgarwal Can you please do a seperate branch / PR for the dashboard stats design updates.
> 
> It will.be easier to review in isolation.
> 
> Plus - we may end up getting the stats thing done much easier due to the discussion about header component stuff that is ongoing.

Sure

lb- commented on wagtail/admin/templates/wagtailadmin/shared/header.html:
```diff
@@ -8,19 +8,24 @@
 
 {% endcomment %}
-<header class="header {% if merged %}merged{% endif %} {% if search_form %}hasform{% endif %}">
+<header class="header {{ classname }} {{ merged|yesno:" merged ," }}{{ search_form|yesno:" hasform ," }}">
```
```suggestion
<header class="header {{ classname }} {{ merged|yesno:" merged ," }}{{ search_form|yesno:" hasform ," }}">
```

thinking long term - it might be nice to have a generic template tag for classnames (similar to https://www.npmjs.com/package/classnames)

```
<header class="{% classnames "header" classname merged|yesno:" merged," search_form|yesno:"hasform," %}">
```

Whatever you give it as args, it will try to parse and spit out a nice classname string without extra spaces.

Give this a go.


----

https://docs.djangoproject.com/en/4.0/howto/custom-template-tags/#django.template.Library.simple_tag

use simple tags, something like...

```python
register = template.Library()

@register.simple_tag
def classnames(*classes):
    #classes # will be an array
    return ' '.join([classname for classname in classes if classname])
```

It will need a test case or two but could come in handy.

gasman replied:
Just curious why we're going to so much trouble here to strip the extra spaces? (Ease of writing test assertions, perhaps?)

lb- replied:
Yeah. It's been a pain in a bunch of previous unit tests.

PaarthAgarwal replied:
I'll try this approach soon. Other than that it's ready for review.

lb- replied:
Thanks, not critical, only go down this path if we get some benefits from it.

PaarthAgarwal replied:
Tried it but It is not working. Getting the same output `class="header  merged"` i.e with double spaces.

lb- replied:
Be careful what you are passing into the template tag. 

```
<header class="header {{ classname }} {{ merged|yesno:" merged ," }}{{ search_form|yesno:" hasform ," }}">

```

This will add a space before the merged string as is. 

But...
```
<header class="header {{ classname }} {{ merged|yesno:"merge ," }}{{ search_form|yesno:" hasform ," }}">
```

Will not. 

Have a read about how yesno works - it will output whatever is inside the supplied string after yesno: (before the comma) if true. 

https://docs.djangoproject.com/en/4.0/ref/templates/builtins/#yesno

lb- replied:
Comment above fixed up - code snippets pasted wrong initially

lb- commented on wagtail/admin/templates/wagtailadmin/shared/header.html:
```diff
@@ -8,19 +8,24 @@
+                        {% icon class_name="w-header__icon" name=icon %}
+                    {% elif avatar %}
+                        <div class="avatar"><img src="{{ avatar }}" alt="" /></div>
+                    {% endif %}
```
This is not really positioned correctly yet, so will need something like 

https://github.com/wagtail/wagtail/blob/ffda5580159094c941053a78fbaea1e1980823fb/client/scss/components/_header.scss#L21-L23

For the `h1 .avatar`

lb- replied:
another approach may be a neg margin.
`margin-inline-start: -2rem;`
(not sure on amount, just guessing)

lb- commented:
@PaarthAgarwal  a few other things

* We can probably remove `header--home` styles as not used now
* Might be worth adding the prefix of `w-header` to the root header class + `w-header__title` etc (but leave left/col/row as is)

PaarthAgarwal commented:
> * w-header

1. Removed the header--home styles.
2. Added the prefix w- to header classes in HTML, but it'll be better to put those changes in a separate PR because it may open a can of worms with failing tests and styling issues.
3. Updated screenshots in the first comment.

lb- commented:
good call on the classes @PaarthAgarwal thanks.

lb- commented on client/scss/layouts/_home.scss:
```diff
@@ -1,30 +1,6 @@
-    font-size: 1.3em;
-    font-weight: 600;
+.homepage {
+  h1 {
```
I recommend we put these styles in the `header`. while there should only be 1 h1 - better to be safe.

We should also use the `header__title` class instead of styling on the `h1` directly.

PaarthAgarwal replied:
done

lb- commented on wagtail/admin/templates/wagtailadmin/shared/header.html:
```diff
@@ -8,19 +8,24 @@
     - `search_url` - if present, display a search box. This is a URL route name (taking no parameters) to be used as the action for that search box
     - `query_parameters` - a query string (without the '?') to be placed after the search URL
     - `icon` - name of an icon to place against the title
+    - `avatar` - if present, display an 'avatar' in place of icon. This is the URL to be used as the img src for avatar 
```
Sorry. Forgot to mention. We should add `classname` here as it will now be a supported variable

lb- commented on wagtail/admin/templates/wagtailadmin/home/branding_welcome_title.html:
```diff
@@ -0,0 +1,4 @@
+{% load i18n %}
+{% block branding_welcome %}
+    {% blocktrans trimmed %}Welcome to the {{ site_name }} Wagtail CMS{% endblocktrans %}
+{% endblock %}
```
might be good to add a new line here and also configure your editor to add line breaks at the end of your file.

lb- commented on wagtail/admin/templatetags/wagtailadmin_tags.py:
```diff
@@ -153,6 +153,12 @@ def page_permissions(context, page):
 
+@register.simple_tag
+def classnames(*classes):
+    # classes # will be an array
```
```suggestion
    """
    Returns any args as a space-separated joined string for using in HTML class names.
    """
```

lb- reviewed (commented):
Code is looking pretty solid.
Now need to do a review of the UI changes in a few browsers.

- would be good to add a bit more margin after the avatar or pull the avatar over to the left a bit more.

lb- commented:
Update - there may be a refined design for the home (dashboard) view coming in the next week or so so this can probably come in mostly as is.

thibaudcolas commented on docs/advanced_topics/customisation/admin_templates.md:
````diff
@@ -75,12 +75,13 @@ To replace the login message, create a template file `dashboard/templates/wagtai
 
-```html+django
-{% extends "wagtailadmin/home.html" %}
+To replace the welcome message on the admin home page (dashboard), create a template file `dashboard/templates/wagtailadmin/home/branding_welcome_title.html` that overrides the content:
````
Note that if we went ahead with an approach like #8802, we wouldn’t need to refactor this as much.

lb- commented:
@PaarthAgarwal  I have done some tweaks to this branch and will merge in shortly - basically applying what we discussed, but also including Thibaud's template approach https://github.com/wagtail/wagtail/pull/8802

lb- commented on client/scss/layouts/_home.scss:
```diff
@@ -1,31 +1,9 @@
+.homepage {
+  header {
+    .header__title {
+      inset-inline-start: -1.5em;
```
To simplify things I opted for a shared header styles approach only, see other comment below.

lb- commented on wagtail/admin/templates/wagtailadmin/shared/header.html:
```diff
@@ -7,20 +7,26 @@
                 <h1 class="header__title">
-                    {% if icon %}{% icon name=icon %}{% endif %}
+                    {% if icon %}
+                        {% icon class_name="w-header__icon" name=icon %}
```
```suggestion
                        {% icon class_name="w-header__glyph" name=icon %}
```

Having a generic class for the icon and avatar - we can make styles a bit easier to follow.

PaarthAgarwal replied:
That'll be great!

lb- commented on wagtail/admin/templates/wagtailadmin/shared/header.html:
```diff
@@ -7,20 +7,26 @@
+                    {% if icon %}
+                        {% icon class_name="w-header__icon" name=icon %}
+                    {% elif avatar %}
+                        <div class="avatar"><img src="{{ avatar }}" alt="" /></div>
```
```suggestion
                        <div class="w-header__glpyh avatar"><img src="{{ avatar }}" alt="" /></div>
```

Similar to above.
